src/idl_gen_js.cpp

```diff
@@ -0,0 +1,712 @@
+// outside of its namespace.
+static std::string WrapInNameSpace(const Namespace *ns,
+                                   const std::string &name) {
+  std::string qualified_name;
+  for (auto it = ns->components.begin();
+           it != ns->components.end(); ++it) {
+    qualified_name += *it + ".";
+  }
+  return qualified_name + name;
+}
+
+static std::string WrapInNameSpace(const Definition &def) {
+  return WrapInNameSpace(def.defined_namespace, def.name);
+}
+
+// Generate a documentation comment, if available.
+static void GenDocComment(const std::vector<std::string> &dc,
+                          std::string *code_ptr,
+                          const std::string &extra_lines,
+                          const char *indent = nullptr) {
+  if (dc.empty() && extra_lines.empty()) {
+    // Don't output empty comment blocks with 0 lines of comment content.
+    return;
+  }
+
+  std::string &code = *code_ptr;
+  if (indent) code += indent;
+  code += "/**\n";
+  for (auto it = dc.begin(); it != dc.end(); ++it) {
+    if (indent) code += indent;
+    code += " *" + *it + "\n";
+  }
+  if (!extra_lines.empty()) {
+    if (!dc.empty()) {
+      if (indent) code += indent;
+      code += " *\n";
+    }
+    if (indent) code += indent;
+    std::string::size_type start = 0;
+    while (true) {
+      auto end = extra_lines.find('\n', start);
+      if (end != std::string::npos) {
+        code += " * " + extra_lines.substr(start, end - start) + "\n";
+        start = end + 1;
+      } else {
+        code += " * " + extra_lines.substr(start) + "\n";
+        break;
+      }
+    }
+  }
+  if (indent) code += indent;
+  code += " */\n";
+}
+
+static void GenDocComment(std::string *code_ptr,
+                          const std::string &extra_lines) {
+  GenDocComment(std::vector<std::string>(), code_ptr, extra_lines);
+}
+
+// Generate an enum declaration and an enum string lookup table.
+static void GenEnum(EnumDef &enum_def, std::string *code_ptr,
+                    std::string *exports_ptr) {
+  if (enum_def.generated) return;
+  std::string &code = *code_ptr;
+  std::string &exports = *exports_ptr;
+  GenDocComment(enum_def.doc_comment, code_ptr, "@enum");
+  if (enum_def.defined_namespace->components.empty()) {
+    code += "var ";
+    exports += "this." + enum_def.name + " = " + enum_def.name + ";\n";
+  }
+  code += WrapInNameSpace(enum_def) + " = {\n";
+  for (auto it = enum_def.vals.vec.begin();
+       it != enum_def.vals.vec.end(); ++it) {
+    auto &ev = **it;
+    if (!ev.doc_comment.empty()) {
+      if (it != enum_def.vals.vec.begin()) {
+        code += '\n';
+      }
+      GenDocComment(ev.doc_comment, code_ptr, "", "  ");
+    }
+    code += "  " + ev.name + ": " + NumToString(ev.value);
+    code += (it + 1) != enum_def.vals.vec.end() ? ",\n" : "\n";
+  }
+  code += "};\n\n";
+}
+
+static std::string GenType(const Type &type) {
+  switch (type.base_type) {
```

Comment:
I guess this works too, though I guess all other languages add these to the table in idl.h, so that you can use a single macro instead of switches to generate these. Fine either way though.